core/node/node_framework/src/service/mod.rs

```diff
@@ -190,13 +197,33 @@ impl ZkStackService {
             .block_on(futures::future::join_all(remaining_tasks_with_timeout));
         let execution_timeouts_count = execution_results.iter().filter(|&r| r.is_err()).count();
         if execution_timeouts_count > 0 {
-            tracing::warn!(
+            tracing::error!(
                 "{execution_timeouts_count} tasks didn't finish in {TASK_SHUTDOWN_TIMEOUT:?} and were dropped"
             );
         } else {
             tracing::info!("Remaining tasks finished without reaching timeouts");
         }
 
+        // Run shutdown hooks sequentially.
+        for hook in shutdown_hooks {
+            let name = hook.id().clone();
+            // Limit each shutdown hook to the same timeout as the tasks.
+            let hook_with_timeout =
+                async move { tokio::time::timeout(TASK_SHUTDOWN_TIMEOUT, hook.invoke()).await };
+            match self.runtime.block_on(hook_with_timeout) {
+                Ok(Ok(())) => {
+                    tracing::info!("Shutdown hook {name} completed");
+                }
+                Ok(Err(err)) => {
+                    tracing::error!("Shutdown hook {name} failed: {err}");
+                    // We still have to invoke all the remaining hooks, so we don't return early.
```

Comment:
Are errors in shutdown hooks not propagated intentionally? Before the node framework, e.g. failing to terminate RocksDB instances was propagated.

Reply:
That's a good point. Initially when I wrote this code I thought "we're exiting anyways, so who cares", but indeed propagating the error would probably be a better approach. I guess I'll do it similarly to how task work -- the first occurred error will be propagated.

Reply:
I think it should make sense to return an error (one of the errors if several are returned, I suppose) provided that the main task logic didn't return an error.

Reply:
Yup, already in works.